File: java/Linked_List/Copy_List_with_Random_pointer.java
package java.Linked_List;

class Node {
  int val;
  Node next;
  Node random;

  Node(int val) {
    this.val = val;
  }
}

public class Copy_List_with_Random_pointer {
  public Node copyRandomList(Node head) {
    if (head == null)
      return null;
    Node current = head;
    while (current != null) {
      Node copy = new Node(current.val);
      copy.next = current.next;
      current.next = copy;
      current = copy.next;
    }
    current = head;
    while (current != null) {
      if (current.random != null)
        current.next.random = current.random.next;
      current = current.next.next;
    }
    current = head;
    Node dummy = new Node(0);
    Node copyCurrent = dummy;
    while (current != null) {
      copyCurrent.next = current.next;
      current.next = current.next.next;
      current = current.next;
      copyCurrent = copyCurrent.next;
    }
    return dummy.next;
  }
}

File: python/Linked_List/Copy_List_withRandom_Pointer.py
class Node:
  def __init__(self, val=0, next=None, random=None):
    self.val = val
    self.next = next
    self.random = random

class Solution:
  def copyRandomList(self, head):
    oldToNew = {None: None}

    curr = head 

    while curr:
      oldToNew[curr] = Node(curr.val)
      curr = curr.next 
    
    curr = head 

    while curr:
      copy = oldToNew[curr]
      copy.next = oldToNew[curr.next]
      copy.random = oldToNew[curr.random]
    
    return oldToNew[head]